traffic_ops/client/server.go

```diff
@@ -33,158 +33,144 @@ const (
 		}
 		if len(st) == 0 {
-			return tc.Alerts{}, ReqInf{}, errors.New("no status named " + server.Status)
+			return alerts, reqInf, fmt.Errorf("no status named %s", *server.Status)
 		}
-		server.StatusID = st[0].ID
+		server.StatusID = &st[0].ID
 	}
-	if server.TypeID == 0 && server.Type != "" {
+	if (server.TypeID == nil || *server.TypeID == 0) && server.Type != "" {
 		ty, _, err := to.GetTypeByName(server.Type)
 		if err != nil {
-			return tc.Alerts{}, ReqInf{}, errors.New("no type named " + server.Type + ":" + err.Error())
+			return alerts, reqInf, fmt.Errorf("no type named %s: %v", server.Type, err)
 		}
 		if len(ty) == 0 {
-			return tc.Alerts{}, ReqInf{}, errors.New("no type named " + server.Type)
+			return alerts, reqInf, fmt.Errorf("no type named %s", server.Type)
 		}
-		server.TypeID = ty[0].ID
+		server.TypeID = &ty[0].ID
 	}
+
 	reqBody, err := json.Marshal(server)
 	if err != nil {
-		return tc.Alerts{}, reqInf, err
+		return alerts, reqInf, err
 	}
 
 	resp, remoteAddr, err := to.request(http.MethodPost, API_SERVERS, reqBody)
+	reqInf.RemoteAddr = remoteAddr
 	if err != nil {
-		return tc.Alerts{}, reqInf, err
+		return alerts, reqInf, err
 	}
 	defer resp.Body.Close()
-	var alerts tc.Alerts
+
 	err = json.NewDecoder(resp.Body).Decode(&alerts)
-	return alerts, reqInf, nil
+	return alerts, reqInf, err
 }
 
 // UpdateServerByID updates a Server by ID.
-func (to *Session) UpdateServerByID(id int, server tc.Server) (tc.Alerts, ReqInf, error) {
-
+func (to *Session) UpdateServerByID(id int, server tc.ServerNullable) (tc.Alerts, ReqInf, error) {
+	var alerts tc.Alerts
 	var remoteAddr net.Addr
-	reqBody, err := json.Marshal(server)
 	reqInf := ReqInf{CacheHitStatus: CacheHitStatusMiss, RemoteAddr: remoteAddr}
+
+	reqBody, err := json.Marshal(server)
 	if err != nil {
-		return tc.Alerts{}, reqInf, err
+		return alerts, reqInf, err
 	}
+
 	route := fmt.Sprintf("%s/%d", API_SERVERS, id)
 	resp, remoteAddr, err := to.request(http.MethodPut, route, reqBody)
+	reqInf.RemoteAddr = remoteAddr
 	if err != nil {
-		return tc.Alerts{}, reqInf, err
+		return alerts, reqInf, err
 	}
 	defer resp.Body.Close()
-	var alerts tc.Alerts
+
 	err = json.NewDecoder(resp.Body).Decode(&alerts)
-	return alerts, reqInf, nil
+	return alerts, reqInf, err
 }
 
 // GetServers returns a list of Servers.
-func (to *Session) GetServers() ([]tc.Server, ReqInf, error) {
-	resp, remoteAddr, err := to.request(http.MethodGet, API_SERVERS, nil)
-	reqInf := ReqInf{CacheHitStatus: CacheHitStatusMiss, RemoteAddr: remoteAddr}
-	if err != nil {
-		return nil, reqInf, err
+// The 'params' parameter can be used to optionally pass URL "query string
+// parameters" in the request.
+// It returns, in order, the requested servers, any Alerts the API returned,
+// the returned summary.count value, a request info object, and any error that
+// occurred.
+func (to *Session) GetServers(params *url.Values) ([]tc.ServerNullable, tc.Alerts, uint64, ReqInf, error) {
```

Comment:
We're starting to get pretty heavy in terms of the number of things these client methods return. Should we just return `ServersV3Response` instead of `[]tc.ServerNullable, tc.Alerts, uint64`, and allow clients to get whatever fields they're interested in from that?

Reply:
That's fine with me, as long as that's consistently what we wanna do with these.